src/components/ActionManager/ActionManager.js

```diff
@@ -0,0 +1,119 @@
+import Cosmos from "@lunie/cosmos-js"
+import config from "src/config"
+import { getSigner } from "./signer"
+import transaction from "./transactionTypes"
+
+import { formatCosmosArguments, convertCurrencyData } from "./cosmosMessages"
+
+class ActionManager {
+  setContext(context) {
+    if (!context) {
+      return
+    }
+    this.context = context
+    this.cosmos = new Cosmos(this.context.url || "", this.context.chainId || "")
+  }
+
+  async simulate(type, transactionProperties) {
+    if (!this.context) {
+      throw Error(`This modal has no context.`)
+    }
+
+    if (!this.context.connected) {
+      throw Error(
+        `Currently not connected to a secure node. Please try again when Lunie has secured a connection.`
+      )
+    }
+
+    const txArguments = formatCosmosArguments(
+      this.context,
+      type,
+      transactionProperties,
+      true
+    )
+
+    if (txArguments === null) {
+      throw Error(
+        `Did not complete transaction simulation. Invalid message type: ${type}.`
+      )
+    }
+
+    const message = this.createCosmosMessage(
+      type,
+      this.context.userAddress,
+      txArguments
+    )
+    const gasEstimate = await message.simulate({
+      memo: transactionProperties.memo
+    })
+    return gasEstimate
+  }
+
+  async send(type, transactionProperties, txMetaData) {
+    if (!this.context) {
+      throw Error(`This modal has no context.`)
+    }
+
+    if (!this.context.connected) {
+      throw Error(
+        `Currently not connected to a secure node. Please try again when Lunie has secured a connection.`
+      )
+    }
+
+    const txArguments = formatCosmosArguments(
+      this.context,
+      type,
+      transactionProperties
+    )
+
+    if (txArguments === null) {
```

Comment:
There is a bunch of code repetition between send and simulate, am I wrong?

Reply:
Not wrong, I'll see about this. 👍